mchf-eclipse/misc/uhsdr_ring_buffer.c

```diff
@@ -0,0 +1,120 @@
+ **                                                                                **
+ **                                        UHSDR                                   **
+ **               a powerful firmware for STM32 based SDR transceivers             **
+ **                                                                                **
+ **  Licence:     GNU GPLv3                                                        **
+ ************************************************************************************/
+
+/**
+ * @file uhsdr_ring_buffer.c
+ * @brief Generalized ring buffer.
+ */
+
+#include <assert.h>
+#include "uhsdr_types.h"
+#include "uhsdr_ring_buffer.h"
+
+
+void uhsdr_ring_buffer_init( uhsdr_ring_buffer_t* const r_buf, void *buf, uint32_t size )
+{
+    assert( r_buf && buf && size );
+
+    r_buf->buf         = (uint8_t *)buf;
+    r_buf->size        = size;
+
+    r_buf->head.index = r_buf->head.shadow_index = 0;
+    r_buf->tail.index = r_buf->tail.shadow_index = 0;
+}
+
+uint32_t uhsdr_ring_buffer_get( uhsdr_ring_buffer_t* const r_buf, uint8_t* data )
+{
+    assert( r_buf && data );
+
+    uint32_t retval = 0;
+
+    uint32_t tail_pair;
+    uint32_t new_tail;
+    uhsdr_ring_buffer_ixds_pair_t* tail_pair_ptr = (uhsdr_ring_buffer_ixds_pair_t*) &tail_pair;
+    do
+    {
+        tail_pair = __LDREXW(( volatile uint32_t* )&r_buf->tail );
+
+        /**
+         * if HEAD's index not equal HEAD's shadow_index, that means some thread obtained
+         * index for putting entry into buffer, but not done yet.
+         */
+        if( r_buf->head.index != r_buf->head.shadow_index )
+        {
+            __CLREX();
+            continue;
+        }
+
+        if ( r_buf->head.index != tail_pair_ptr->index )
+        {
+            *data = r_buf->buf[ tail_pair_ptr->index ];
+            new_tail = ( tail_pair_ptr->index + 1 ) %  r_buf->size;
+            retval = 1;
+        }
+        else
+        {
+            /** there is no room, write back old values */
+            new_tail = (uint32_t)tail_pair;
+        }
+    } while( __STREXW( new_tail, ( volatile uint32_t* )&r_buf->tail));
+    __DMB();
+
+    return retval;
+}
+
+uint32_t uhsdr_ring_buffer_put( uhsdr_ring_buffer_t* const r_buf, uint8_t data )
+{
+    assert( r_buf );
+
+    uint32_t retval = 0;
+
+    uint32_t head_pair;
+    uint32_t next_head;
+    uhsdr_ring_buffer_ixds_pair_t* head_pair_ptr = (uhsdr_ring_buffer_ixds_pair_t*) &head_pair;
+    do
+    {
+        head_pair = __LDREXW(( volatile uint32_t* )&r_buf->head );
+        next_head = ( head_pair_ptr->index + 1 ) % r_buf->size;
+        if ( next_head != r_buf->tail.index )
+        {
+            /** There is room */
+            retval = 1;
+        }
+        else
+        {
+            /** There is no room, write back old values */
+            next_head = (uint32_t)head_pair;
+        }
+    } while( __STREXW( next_head, ( volatile uint32_t* )&r_buf->head ));
+
+    /**
```

Comment:
What happens if another uhsdr_ring_buffer_put is executed right here (i.e. in an interrupt of higher priority)?

Reply:
it would use the next index as we already stored...
Let's assume we have to threads, main and an int.(int with higher priority)...
main execute  uhsdr_ring_buffer_put() ( tail = 20, head 30)
it read taill 20, compute next_tail 21, checking that we have room in buffer, stored 21 back and leave while loop... 

and here I probably need add memory barrier as it's the place where the changes should be visible for another observer.

Ok, we stored index. if it prevents by another thread here it would just use 31 index. 
Do not see an issue here.  If you see, could you share?

Reply:
I see issue when we have two threads. P1 and P2. And P1 trying put, prevented by P2 here, P2 put into buffer, P2 continuet execution and later read from buffer. Here the issue, if in buffer just two entries( P1 is not finished puting ) it would read some data, but not what it should be there... 

Need think about it.

Reply:
Our low prio put enters with tail.idx/shdw = 20, head.idx/shdw = 30, so once we leave the guarded section we have our next_head at 31 and would have to store our data at 31, behind the last used slot. With __STREXW we store the next head in the head.idx   (which now reads 31, head.shdw == 30) 
Now we get interrupted with another put. Since we have left the guarded section, the higher prio put will have no trouble to execute. Meaning it will calculate to use slot 32 (since we have head.idx == 31 that this holds data) for storing, then stores the data here and finally advances head.shdw to 32.

And now we in the lower prio put get the control back. We execute the remainder of the code, retval is != 0, so we put our data in  r_buf->buf[ head_pair_ptr->index ] = data, which unfortunately is now pointing at 32. And we'll write our data there over the previously written data, but nothing has been written to slot 31. Which is bad.

I might be wrong, but it seems this algorithm does not fulfill is promise to be reentrant.
 
This hole subject is very, very difficult, that is why I try to stick to proven, well-known "text book" algorithms.

Reply:
> Our low prio put enters with tail.idx/shdw = 20, head.idx/shdw = 30, so once we leave the guarded section we have our next_head at 31 and would have to store our data at 31, behind the last used slot. With __STREXW we store the next head in the head.idx (which now reads 31, head.shdw == 30)
> Now we get interrupted with another put. Since we have left the guarded section, the higher prio put will have no trouble to execute. Meaning it will calculate to use slot 32 (since we have head.idx == 31 that this holds data) for storing, then stores the data here and finally advances head.shdw to 32.
> 
> And now we in the lower prio put get the control back. We execute the remainder of the code, retval is != 0, so we put our data in r_buf->buf[ head_pair_ptr->index ] = data, which unfortunately is now pointing at 32. 

here head_pair_ptr->index still pointed at 31 (head_pair is located on the stuck of the prevented thread)...

If fallow books we should simply play with disabling and enabling irqs....

Reply:
You're right, I was confused a little by the creative use of casts to fake the pair data structure onto a single uin32_t. Nevermind, so this means we store the data at the "correct" location AFTER we stored data behind this location. Meaning we have for a certain period of time incorrect data in the buffer which cannot be detected by a get if it just hits at the right  moment, which could get the data from index 31 before it has been actually written.

Reply:
the get would wait in loop until index and index_shodow would be equal. That's why, if put prevented by one put or one get it's fine, but if prevented by serial of put - get, get will read not stored et data. 

Now we have an unknown for users bug, that both main/an int routines put data into the buffer. So, seems that this approach solves this particular bug. Yes, it's not ideal, it's, as you said, hard to write 100% multi-thread, lock-free, re-entrant ring buffer... 

I would appreciate if you find time to read more carefully and analyze... Hope you find something that I do not see.

It's not rush.

BTW, do you know places where we use ring buffers? replace them by general one later?

Reply:
Got an idea how to fix, will reopen later.